### ISSUE TITLE
feature: User installable services for use by other Plugins

### ISSUE THREAD

**Issue by joelkoz** [SignalK/signalk-server-node#893]
While thinking about the way to add a common persistence mechanism for use by other plugins, it occurred to me it may be useful to have that service, as well as many other types, be ones that can be installed via the App store plugin mechanism just like other plugins.  Posting this idea here for feedback before working on a PR.  The general idea would be:

1. The server plugin API would have an additional method added to it so general purpose plugins could use it.  It would be `getService(serviceName)` or similar.  That would return an object configured an immediately usable by the plugin that requested it.  So, for example, a plugin calling `getService('mongo')` would return an API for the Mongo DB database.  If no such service is available, null would be returned.

2. Plugins installed by the end user would optionally define two additional methods:
    1. `getServiceName()` if defined by the plugin, would indicate the plugin can provide a service API for other plugins.

    2. `getService(plugin, app)` would be a service factory method that would create a service instance for use by the specified plugin.  An optional approach to returning a service interface object would be for `getService()` to simply decorate the `app` object (which is assumed to be the current API instance created for use by the specified plugin) with additional methods (similar to a polyfill).  This second technique would allow for expanding the plugin API via the plugin mechanism vs. app server updates.

By having these services be pluggable, it would allow for different systems with different capabilities to take advantage of the best software stack combination.  For example, there could be different plugins that might be capable of creating a 'mongo' service.  Those would include:

1. One that uses the NeDB project
2. One that uses the TingoDB project
3. One that returns an interface to an actual MongoDB service running on an external machine, possibly even one that is cloud based.

We could use this mechanism to come up with common interfaces, like a 'sql' service that is backed by MySQL, Postgres, sqlite, or any other db that the end user chooses.

**Comment by tkurki** [comment]
I think we are already headed towards a version management nightmare with plugins and without a very carefully thought and implemented version compatibility mechanism this would make things even more complex. What if plugin A needs version 3 of something provided by plugin C and B needs version 2? 

One of the unappreciated strengths of JavaScript is that different versions of a dependency can peacefully coexist - I think it is much easier for a plugin to bring its own dependencies, with internal versioning.

When I talked about extending the api provided to plugins with a way to store relational data I was thinking about just that: providing one standard server-sanctioned storage model for more than trivial amounts of data. But it may very well be that even that is not needed. Also I don't fancy bringing one more persistence abstraction to this world - there's plenty around, none of them perfect.

Before adding this much complexity I think we need a lot more evidence as to this being a problem worth solving in this way.

Also anything more serious, like Postgres, Timescale, MySQL, InfluxDb, MongoDB  etc can not be easily provisioned by the server - unlike sqlite they are separate OS level applications. If this is what we need in the future I would rather explore Docker & Docker compose for provisioning the server and related services and let plugins talk to their backends directly, not through the server or via other plugins.

**Comment by joelkoz** [comment]
> One of the unappreciated strengths of JavaScript is that different versions of a dependency can peacefully coexist - I think it is much easier for a plugin to bring its own dependencies, with internal versioning.

True, but it can also be one of its weaknesses.  What if Plugin A, B, C, and D all need the same version of the same module, and that module happens to have some heft to it (as might be the case with a sophisticated persistence engine).  Now, you've got the same code being included four times.

My thought with this was that NeDB, and TingoDB both emulate the Mongo interface exactly.  So, we wouldn't really be creating another abstraction layer (at least, not in the case).  However, NeDB is an "in memory only".  That may be fine for some installations, but for ones that make heavy use of plugins that make heavy use of the persistence, we could swap out NeDB for one of the others.

That being said, I can't argue strongly enough for this approach right now, and it may very well be over-engineering the problem.  A single common persistence engine may in fact solve the problem of duplicated large modules, and I don't know if there is any other use case right now for this proposed solution.

One thought, however: there is some merit to supporting a mechanism where a plugin can polyfill API methods into the app object passed to other plugins.  This might actually be a good way to deal with versioning problems.  If this mechanism were to be added as an undocumented feature in 1.19, and then at some point in the future, Node Server 2.0 comes out, and  new plugin is created that needs some of the new API call, this mechanism would allow for a "compatibility plugin" to be added to a 1.19 system that polyfills the necessary methods to get 2.0 to work.  Either that, or this imagined plugin could be released with a polyfill plugin without having to wait on the 2.0 release.

Anyway - its worth filing away as a possible solution to future compatibility problems.
